### PR TITLE
Handle internal reddit links internally, including x-posts

### PR DESCRIPTION
**Pull request by obosob**
Thought it made sense to handle crossposts internally too. generally improved the determination of what is a self/cross post.

### CODE REVIEW

File: rtv/subreddit.py [hunk 1]
```diff
@@ -108,10 +108,11 @@ def open_link(self):
         "Open a link with the webbrowser"
         data = self.content.get(self.nav.absolute_index)
 
-        if data['url'] == 'selfpost':
-            self.open_submission()
+        url = data['url_full']
+        if data['url'] != url:
```
Comment (Brobin) [inline rtv/subreddit.py]: This isn't very descriptive as to what's happening. If a cross post's url was represented as simply `'x-post'` instead of `'x-post via /r/{sub}'`, you could do something like this.

```python
if data['url'] in ['selfpost', 'x-post']:
    ...
```

I think this would make things more explicit, and obvious as to what's actually happening.

File: rtv/content.py [hunk 2]
```diff
@@ -123,7 +125,14 @@ def strip_praw_submission(sub):
         data['subreddit'] = strip_subreddit_url(sub.permalink)
         data['flair'] = flair
         data['url_full'] = sub.url
-        data['url'] = ('selfpost' if is_selfpost(sub.url) else sub.url)
+
+        if reddit_link.match(sub.url):
+            stripped_url = reddit_link_no_host.search(sub.url).group()
+            stripped_comments = reddit_link_no_host.search(sub.permalink).group()
+            data['url'] = ('selfpost' if stripped_url == stripped_comments
+                    else 'x-post via {}'.format(strip_subreddit_url(sub.url)) )
```
Comment (Brobin) [inline rtv/content.py]: See previous comment. Would it be better to just have x-posts be `'x-post'`?

**Comment by obosob** [comment]
I just felt it made a better view in the list if you know what sub its x-posted to. maybe i could add that as a special case in draw though. what do you think? I do prefer the readability of your suggestion.

**Comment by Brobin** [comment]
Maybe we could add another key to the data. `data['x-post'] = '/r/python'`, then you could check that for display.

```python
if data['x-post']:
    # display "x-post via /r/sub"
```

**Comment by Brobin** [comment]
Looks great! :+1:

File: rtv/content.py [hunk 3]
```diff
@@ -105,7 +106,8 @@ def strip_praw_submission(sub):
         displayed through the terminal.
         """
 
-        is_selfpost = lambda s: s.startswith('http://www.reddit.com/r/')
+        reddit_link = re.compile("https?://(www.)?(np.)?redd(it.com|.it)/r/.*")
```
Comment (michael-lazar) [inline rtv/content.py]: I'm pretty sure your periods need to be surrounded with ``[.]``, otherwise they act like wildcards.

File: rtv/content.py [hunk 4]
```diff
@@ -154,7 +163,7 @@ def from_url(cls, reddit, url, loader, indent_size=2, max_indent_level=8):
 
         try:
             with loader():
-                submission = reddit.get_submission(url, comment_sort='hot')
+                submission = reddit.get_submission(url.replace("http:","https:"), comment_sort='hot')
```
Comment (michael-lazar) [inline rtv/content.py]: Why is this necessary? If the URL is provided as ``http:``, it would make sense to honor that and let Reddit redirect us to the right location.

Comment (obosob) [inline rtv/content.py]: praw throws an unexpected redirect at this point otherwise. if you know of a better way to handle it then let me know.

File: rtv/content.py (hunk 3 above)
Comment (obosob) [inline rtv/content.py]: damnit, you're right, well, escaping it will be fine, but yep.

File: rtv/content.py [hunk 5]
```diff
@@ -123,7 +125,14 @@ def strip_praw_submission(sub):
...
         data['url_full'] = sub.url
-        data['url'] = ('selfpost' if is_selfpost(sub.url) else sub.url)
+
+        if reddit_link.match(sub.url):
+            stripped_url = reddit_link_no_host.search(sub.url).group()
+            stripped_comments = reddit_link_no_host.search(sub.permalink).group()
+            data['url'] = ('selfpost' if stripped_url == stripped_comments
+                    else 'x-post')
+        else:
+            data['url'] = sub.url
```
Comment (michael-lazar) [inline rtv/content.py]: I think that this the logic here is a kind of confusing to follow. I was thinking of ways to potentially clean it up and came up with this -

The values that we have are:
``data['permalink']`` - Link to the submission comments
``data['url_full']`` - Link that the submission points to
``data['url']`` - What is displayed on the screen

So in keeping with this convention, we should really add a new field for the post type. That way you don't have to hijack the ``data['url']`` for x-posts. I think Brobin was eluding to this also.

```python
if data['permalink'].split('/r/')[-1] == data['url_full'].split('/r/')[-1]:
    data['url_type'] = 'selfpost'
    data['url'] = 'selfpost'

elif reddit_link.match(data['url_full']):
    data['url_type'] = 'x-post'
    data['url'] = 'x-post via {}'.format(data['subreddit'])

else:
    data['url_type'] = 'external'
    data['url'] = data['url_full']
```

This also simplifies subreddit.py because you can go back to printing ``data['url']`` directly.

File: rtv/content.py (hunk 4 above)
Comment (michael-lazar) [inline rtv/content.py]: Makes sense, thanks

File: rtv/content.py (hunk 5 above)
Comment (Brobin) [inline rtv/content.py]: I think this is a great way to  handle this. Having `data['url_type']` instead of combinations of other conditions will make the code much more readable. :+1:

Comment (obosob) [inline rtv/content.py]: +1

**Comment by obosob** [comment]
Bump...

**Comment by michael-lazar** [comment]
Sorry, I must have missed your latest changes. Thanks for the PR. :smiley:
